<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>interminebio.SequenceQuery</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="interminebio-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://www.intermine.org">InterMine Python Webservice Client</a></th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="interminebio-module.html">Package&nbsp;interminebio</a> ::
        Class&nbsp;SequenceQuery
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="interminebio.SequenceQuery-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class SequenceQuery</h1><p class="nomargin-top"><span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery">source&nbsp;code</a></span></p>
<pre class="base-tree">
       object --+    
                |    
<a href="interminebio.SequenceDataQuery-class.html">SequenceDataQuery</a> --+
                    |
                   <strong class="uidshort">SequenceQuery</strong>
</pre>

<hr />
<h1 class="heading">Class for querying InterMine Webservices for Sequence based data</h1>
    <p>This module allows you to construct queries that retrieve data about
    sequences and sequence features in biologically relevant formats.</p>
    <p>The currently supported formats are UCSC-BED, GFF3, and FASTA.</p>

<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="interminebio.SequenceQuery-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">service_or_query</span>,
        <span class="summary-sig-arg">root</span>=<span class="summary-sig-default">None</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="interminebio.SequenceQuery-class.html#add_sequence_feature" class="summary-sig-name">add_sequence_feature</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">*features</span>)</span><br />
      Fasta, GFF3 and BED queries all can read information from 
      SequenceFeatures.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.add_sequence_feature">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="where"></a><span class="summary-sig-name">where</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">*args</span>,
        <span class="summary-sig-arg">**kwargs</span>)</span><br />
      Add a constraint to the query, and return self for chaining.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.where">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="interminebio.SequenceQuery-class.html#set_sequence" class="summary-sig-name">set_sequence</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">f</span>)</span><br />
      Add a sequence holding object to the query.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.set_sequence">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code><a href="interminebio.SequenceDataQuery-class.html">SequenceDataQuery</a></code></b>:
      <code><a href="interminebio.SequenceDataQuery-class.html#bed">bed</a></code>,
      <code><a href="interminebio.SequenceDataQuery-class.html#fasta">fasta</a></code>,
      <code><a href="interminebio.SequenceDataQuery-class.html#gff3">gff3</a></code>
      </p>
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__repr__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">service_or_query</span>,
        <span class="sig-arg">root</span>=<span class="sig-default">None</span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.__init__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <h1 class="heading">Constructor</h1>
<pre class="py-doctest">
<span class="py-prompt">&gt;&gt;&gt; </span>s = Service(<span class="py-string">&quot;www.flymine.org/query&quot;</span>)
<span class="py-prompt">&gt;&gt;&gt; </span>bio_query = SequenceQuery(s, <span class="py-string">&quot;Gene&quot;</span>)
<span class="py-output">&lt;interminebio.SequenceQuery xxx&gt;</span>
<span class="py-output"></span><span class="py-prompt">&gt;&gt;&gt; </span>q = s.new_query(<span class="py-string">&quot;Gene&quot;</span>).where(s.model.Gene.symbol == [<span class="py-string">&quot;h&quot;</span>, <span class="py-string">&quot;r&quot;</span>, <span class="py-string">&quot;eve&quot;</span>, <span class="py-string">&quot;zen&quot;</span>])
<span class="py-prompt">&gt;&gt;&gt; </span>bio_query = SequenceQuery(q)
<span class="py-output">&lt;interminebio.SequenceQuery yyy&gt;</span></pre>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>service_or_query</code></strong> (intermine.webservice.Service or intermine.query.Query) - The service to connect to, or a query to wrap.</li>
        <li><strong class="pname"><code>root</code></strong> (str) - The root class of the query</li>
    </ul></dd>
    <dt>Overrides:
        object.__init__
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="add_sequence_feature"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">add_sequence_feature</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">*features</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.add_sequence_feature">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <h1 class="heading">Add an arbitrarily long list of sequence features to the query.</h1>
    <p>Fasta, GFF3 and BED queries all can read information from 
    SequenceFeatures. For Fasta you are advised to use the set_sequence 
    method instead, as unlike the GFF3 and BED services, the Fasta service 
    can only handle queries with one output column.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="set_sequence"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">set_sequence</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">f</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="interminebio-pysrc.html#SequenceQuery.set_sequence">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <h1 class="heading">Set the sequence column to retrieve.</h1>
    <p>Add a sequence holding object to the query. It can be a 
    SequenceFeature, Protein or Sequence object.</p>
    <p>Fasta queries, which read sequences rather than sequence features, 
    currently only permit one output column.</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="interminebio-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            ><a class="navbar" target="_top" href="http://www.intermine.org">InterMine Python Webservice Client</a></th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Thu Oct  6 11:59:21 2011
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
