--- autointerface.py.orig	2010-07-26 13:31:21.000000000 -0400
+++ autointerface.py	2010-12-18 12:48:48.702921613 -0500
@@ -8,6 +8,7 @@
     from sphinx.domains.python import PyClasslike
 from sphinx.ext import autodoc
 from zope.interface import Interface
+from zope.interface.interface import InterfaceClass
 
 class InterfaceDesc(PyClasslike):
     def get_index_text(self, modname, name_cls):
@@ -29,8 +30,7 @@
 
     @classmethod
     def can_document_member(cls, member, membername, isattr, parent):
-        return isinstance(member, types.ClassType) and \
-                issubclass(member, Interface)
+        return isinstance(member, InterfaceClass)
 
     def add_directive_header(self, sig):
         if self.doc_as_attr:
