

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Phylogenetic Trees &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.1 beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="The ETE tutorial" href="index.html" />
    <link rel="next" title="Clustering Trees" href="tutorial_clustering.html" />
    <link rel="prev" title="The Programmable Tree Drawing Engine" href="tutorial_drawing.html" />
 

<!-- <div class="warning" align="center"> -->
<!--   This site is still under construction and refers to -->
<!--   the new <b>ETE 2.1</b> version (temporary module -->
<!--   name: <b>ete2a1</b>). Many sections may be incomplete or contain -->
<!--   errors. -->
<!-- </div> -->


<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-783841-12']);
  _gaq.push(['_setCookiePath', '/ete2/']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>






  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_clustering.html" title="Clustering Trees"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_drawing.html" title="The Programmable Tree Drawing Engine"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="phylogenetic-trees">
<h1><a class="toc-backref" href="#id1">Phylogenetic Trees</a><a class="headerlink" href="#phylogenetic-trees" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#phylogenetic-trees" id="id1">Phylogenetic Trees</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments" id="id3">Linking Phylogenetic Trees with Multiple Sequence Alignments</a></li>
<li><a class="reference internal" href="#visualization-of-phylogenetic-trees" id="id4">Visualization of phylogenetic trees</a></li>
<li><a class="reference internal" href="#adding-taxonomic-information" id="id5">Adding taxonomic information</a><ul>
<li><a class="reference internal" href="#automatic-control-of-species-info" id="id6">Automatic control of species info</a></li>
<li><a class="reference internal" href="#automatic-and-custom-control-of-the-species-info" id="id7">Automatic (and custom) control of the species info</a></li>
<li><a class="reference internal" href="#manual-control-of-the-species-info" id="id8">Manual control of the species info</a></li>
</ul>
</li>
<li><a class="reference internal" href="#detecting-evolutionary-events" id="id9">Detecting evolutionary events</a><ul>
<li><a class="reference internal" href="#species-overlap-so-algorithm" id="id10">Species Overlap (SO) algorithm</a></li>
<li><a class="reference internal" href="#tree-reconciliation-algorithm" id="id11">Tree reconciliation algorithm</a></li>
<li><a class="reference internal" href="#a-closer-look-to-the-evolutionary-event-object" id="id12">A closer look to the evolutionary event object</a></li>
</ul>
</li>
<li><a class="reference internal" href="#relative-dating-phylogenetic-nodes" id="id13">Relative dating phylogenetic nodes</a><ul>
<li><a class="reference internal" href="#implementation" id="id14">Implementation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#automatic-rooting-outgroup-detection" id="id15">Automatic rooting (outgroup detection)</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Phylogenetic trees are the result of most evolutionary analyses. They
represent the evolutionary relationships among a set of species or, in
molecular biology, a set of homologous sequences.</p>
<p>The <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> class is an extension of the base <a class="reference internal" href="../reference/reference_tree.html#ete2.Tree" title="ete2.Tree"><tt class="xref py py-class docutils literal"><span class="pre">Tree</span></tt></a>
object, providing a appropriate way to deal with phylogenetic trees.
Thus, while leaves are considered to represent species (or sequences
from a given species genome), internal nodes are considered ancestral
nodes. A direct consequence of this is, for instance, that every split
in the tree will represent a speciation or duplication event.</p>
</div>
<div class="section" id="linking-phylogenetic-trees-with-multiple-sequence-alignments">
<h2><a class="toc-backref" href="#id3">Linking Phylogenetic Trees with Multiple Sequence Alignments</a><a class="headerlink" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> instances allow molecular phylogenies to be linked
to the Multiple Sequence Alignments (MSA). To associate a MSA with a
phylogenetic tree you can use the <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.link_to_alignment" title="ete2.PhyloNode.link_to_alignment"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.link_to_alignment()</span></tt></a>
method. You can use the <tt class="xref py py-attr docutils literal"><span class="pre">alg_format</span></tt> argument to specify its
format (See <a class="reference internal" href="../reference/reference_seqgroup.html#ete2.SeqGroup" title="ete2.SeqGroup"><tt class="xref py py-class docutils literal"><span class="pre">SeqGroup</span></tt></a> documentation for available formats)</p>
<p>Given that Fasta format are not only applicable for MSA but also for
<strong>Unaligned Sequences</strong>, you may also associate sequences of different
lengths with tree nodes.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">fasta_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">&gt;seqA</span>
<span class="s">MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s">&gt;seqB</span>
<span class="s">MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s">&gt;seqC</span>
<span class="s">MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s">&gt;seqD</span>
<span class="s">MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s">&quot;&quot;&quot;</span>

<span class="c"># Load a tree and link it to an alignment.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">)</span>
<span class="n">t</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>The same could be done at the same time the tree is being loaded, by
using the <tt class="xref py py-attr docutils literal"><span class="pre">alignment</span></tt> and <tt class="xref py py-attr docutils literal"><span class="pre">alg_format</span></tt> arguments of
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Load a tree and link it to an alignment.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As currently implemented, sequence linking process is not strict,
which means that a perfect match between all node names and sequences
names <strong>is not required</strong>. Thus, if only one match is found between
sequences names within the MSA file and tree node names, only one tree
node will contain an associated sequence. Also, it is important to
note that sequence linking is not limited to terminal nodes. If
internal nodes are named, and such names find a match within the
provided MSA file, their corresponding sequences will be also loaded
into the tree structure. Once a MSA is linked, sequences will be
available for every tree node through its <tt class="xref py py-attr docutils literal"><span class="pre">node.sequence</span></tt>
attribute.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="n">fasta_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> &gt;seqA</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;seqB</span>
<span class="s"> MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s"> &gt;seqC</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;seqD</span>
<span class="s"> MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">iphylip_txt</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> 4 76</span>
<span class="s">      seqA   MAEIPDETIQ QFMALT---H NIAVQYLSEF GDLNEALNSY YASQTDDIKD RREEAHQFMA</span>
<span class="s">      seqB   MAEIPDATIQ QFMALTNVSH NIAVQY--EF GDLNEALNSY YAYQTDDQKD RREEAHQFMA</span>
<span class="s">      seqC   MAEIPDATIQ ---ALTNVSH NIAVQYLSEF GDLNEALNSY YASQTDDQPD RREEAHQFMA</span>
<span class="s">      seqD   MAEAPDETIQ QFMALTNVSH NIAVQYLSEF GDLNEAL--- ---------- -REEAHQ---</span>
<span class="s">             LTNVSHQFMA LTNVSH</span>
<span class="s">             LTNVSH---- ------</span>
<span class="s">             LTNVSH---- ------</span>
<span class="s">             -------FMA LTNVSH</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="c"># Load a tree and link it to an alignment. As usual, &#39;alignment&#39; can</span>
<span class="c"># be the path to a file or data in text format.</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((seqA,seqB),seqC),seqD);&quot;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="n">fasta_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>

<span class="c">#We can now access the sequence of every leaf node</span>
<span class="k">print</span> <span class="s">&quot;These are the nodes and its sequences:&quot;</span>
<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span>
<span class="c">#seqD MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="c">#seqC MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="c">#seqA MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="c">#seqB MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="c">#</span>
<span class="c"># The associated alignment can be changed at any time</span>
<span class="n">t</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="o">=</span><span class="n">iphylip_txt</span><span class="p">,</span> <span class="n">alg_format</span><span class="o">=</span><span class="s">&quot;iphylip&quot;</span><span class="p">)</span>
<span class="c"># Let&#39;s check that sequences have changed</span>
<span class="k">print</span> <span class="s">&quot;These are the nodes and its re-linked sequences:&quot;</span>
<span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">leaf</span><span class="o">.</span><span class="n">sequence</span>
<span class="c">#seqD MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAHQ----------FMALTNVSH</span>
<span class="c">#seqC MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAHQFMALTNVSH----------</span>
<span class="c">#seqA MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQFMALTNVSHQFMALTNVSH</span>
<span class="c">#seqB MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAHQFMALTNVSH----------</span>
<span class="c">#</span>
<span class="c"># The sequence attribute is considered as node feature, so you can</span>
<span class="c"># even include sequences in your extended newick format!</span>
<span class="k">print</span> <span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;sequence&quot;</span><span class="p">],</span> <span class="n">format</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c"># (((seqA[&amp;&amp;NHX:sequence=MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQF</span>
<span class="c"># MALTNVSHQFMALTNVSH],seqB[&amp;&amp;NHX:sequence=MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSY</span>
<span class="c"># YAYQTDDQKDRREEAHQFMALTNVSH----------]),seqC[&amp;&amp;NHX:sequence=MAEIPDATIQ---ALTNVSHNIA</span>
<span class="c"># VQYLSEFGDLNEALNSYYASQTDDQPDRREEAHQFMALTNVSH----------]),seqD[&amp;&amp;NHX:sequence=MAEAPD</span>
<span class="c"># ETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAHQ----------FMALTNVSH]);</span>
<span class="c">#</span>
<span class="c"># And yes, you can save this newick text and reload it into a PhyloTree instance.</span>
<span class="n">sametree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;sequence&quot;</span><span class="p">]))</span>
<span class="k">print</span> <span class="s">&quot;Recovered tree with sequence features:&quot;</span>
<span class="k">print</span> <span class="n">sametree</span>
<span class="c">#</span>
<span class="c">#                              /-seqA</span>
<span class="c">#                    /--------|</span>
<span class="c">#          /--------|          \-seqB</span>
<span class="c">#         |         |</span>
<span class="c">#---------|          \-seqC</span>
<span class="c">#         |</span>
<span class="c">#          \-seqD</span>
<span class="c">#</span>
<span class="k">print</span> <span class="s">&quot;seqA sequence:&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">t</span><span class="o">&amp;</span><span class="s">&quot;seqA&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sequence</span>
<span class="c"># MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAHQFMALTNVSHQFMALTNVSH</span>
</pre></div>
</div>
</div>
<div class="section" id="visualization-of-phylogenetic-trees">
<h2><a class="toc-backref" href="#id4">Visualization of phylogenetic trees</a><a class="headerlink" href="#visualization-of-phylogenetic-trees" title="Permalink to this headline">¶</a></h2>
<p>PhyloTree instances can benefit from all the features of the
programmable drawing engine. However, a built-in phylogenetic layout
is provided for convenience.</p>
<p>All PhyloTree instances are, by default, attached to such layout for
tree visualization, thus allowing for in-place alignment visualization
and evolutionary events labeling.</p>
<div class="figure">
<img alt="../_images/phylotree.png" src="../_images/phylotree.png" />
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>

<span class="n">alg</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s"> &gt;Dme_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEAL--YYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Dme_002</span>
<span class="s"> MAEIPDATIQQFMALTNVSHNIAVQY--EFGDLNEALNSYYAYQTDDQKDRREEAH</span>
<span class="s"> &gt;Cfa_001</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;Mms_001</span>
<span class="s"> MAEAPDETIQQFMALTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s"> &gt;Hsa_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Ptr_002</span>
<span class="s"> MAEIPDATIQ-FMALTNVSHNIAVQY--EFGDLNEALNSY--YQTDDQKDRREEAH</span>
<span class="s"> &gt;Mmu_002</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s"> &gt;Hsa_002</span>
<span class="s"> MAEAPDETIQQFM-LTNVSHNIAVQYLSEFGDLNEAL--------------REEAH</span>
<span class="s"> &gt;Mmu_001</span>
<span class="s"> MAEIPDETIQQFMALT---HNIAVQYLSEFGDLNEALNSYYASQTDDIKDRREEAH</span>
<span class="s"> &gt;Ptr_001</span>
<span class="s"> MAEIPDATIQ-FMALTNVSHNIAVQY--EFGDLNEALNSY--YQTDDQKDRREEAH</span>
<span class="s"> &gt;Mmu_001</span>
<span class="s"> MAEIPDATIQ---ALTNVSHNIAVQYLSEFGDLNEALNSYYASQTDDQPDRREEAH</span>
<span class="s">&quot;&quot;&quot;</span>

<span class="c"># Performs a tree reconciliation analysis </span>
<span class="n">gene_tree_nw</span> <span class="o">=</span> <span class="s">&#39;((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),(Ptr_002,(Hsa_002,Mmu_002))));&#39;</span>
<span class="n">species_tree_nw</span> <span class="o">=</span> <span class="s">&quot;((((Hsa, Ptr), Mmu), (Mms, Cfa)), Dme);&quot;</span>
<span class="n">genetree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">gene_tree_nw</span><span class="p">)</span>
<span class="n">sptree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">species_tree_nw</span><span class="p">)</span>
<span class="n">recon_tree</span><span class="p">,</span> <span class="n">events</span> <span class="o">=</span> <span class="n">genetree</span><span class="o">.</span><span class="n">reconcile</span><span class="p">(</span><span class="n">sptree</span><span class="p">)</span>
<span class="n">recon_tree</span><span class="o">.</span><span class="n">link_to_alignment</span><span class="p">(</span><span class="n">alg</span><span class="p">)</span>

<span class="c"># Visualize the reconciled tree</span>
<span class="n">recon_tree</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="s">&quot;phylotree.png&quot;</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">750</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-taxonomic-information">
<h2><a class="toc-backref" href="#id5">Adding taxonomic information</a><a class="headerlink" href="#adding-taxonomic-information" title="Permalink to this headline">¶</a></h2>
<p id="taxonomic-info"><a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> instances allow to deal with leaf names and species
names separately.  This is useful when working with molecular
phylogenies, in which node names usually represent sequence
identifiers.  Species names will be stored in the
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.species" title="ete2.PhyloNode.species"><tt class="xref py py-attr docutils literal"><span class="pre">PhyloNode.species</span></tt></a> attribute of each leaf node. The method
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_species" title="ete2.PhyloNode.get_species"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_species()</span></tt></a> can be used obtain the set of species
names found under a given internal node (speciation or duplication
event).  Often, sequence names do contain species information as a
part of the name, and ETE can parse this information automatically.</p>
<p>There are three ways to establish the species of the different tree
nodes:</p>
<ul class="simple">
<li>Default: The three first letters of node&#8217;s name represent the species</li>
<li>The species code of each node is dynamically created based on node&#8217;s name</li>
<li>The species code of each node is manually set.</li>
</ul>
<div class="section" id="automatic-control-of-species-info">
<h3><a class="toc-backref" href="#id6">Automatic control of species info</a><a class="headerlink" href="#automatic-control-of-species-info" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree (using default species name encoding)</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>
<span class="c">#                              /-Hsa_001</span>
<span class="c">#                    /--------|</span>
<span class="c">#                   |          \-Ptr_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |         |          /-Cfa_001</span>
<span class="c">#         |          \--------|</span>
<span class="c">#---------|                    \-Mms_001</span>
<span class="c">#         |</span>
<span class="c">#         |          /-Dme_001</span>
<span class="c">#          \--------|</span>
<span class="c">#                    \-Dme_002</span>
<span class="c">#</span>
<span class="c"># Prints current leaf names and species codes</span>
<span class="k">print</span> <span class="s">&quot;Deafult mode:&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>
<span class="c"># node: Dme_001 Species name: Dme</span>
<span class="c"># node: Dme_002 Species name: Dme</span>
<span class="c"># node: Hsa_001 Species name: Hsa</span>
<span class="c"># node: Ptr_001 Species name: Ptr</span>
<span class="c"># node: Cfa_001 Species name: Cfa</span>
<span class="c"># node: Mms_001 Species name: Mms</span>
</pre></div>
</div>
</div>
<div class="section" id="automatic-and-custom-control-of-the-species-info">
<h3><a class="toc-backref" href="#id7">Automatic (and custom) control of the species info</a><a class="headerlink" href="#automatic-and-custom-control-of-the-species-info" title="Permalink to this headline">¶</a></h3>
<p>The default behavior can be changed by using the
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.set_species_naming_function" title="ete2.PhyloNode.set_species_naming_function"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></tt></a> method or by using the
<tt class="xref py py-attr docutils literal"><span class="pre">sp_naming_function</span></tt> argument of the <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> class.
Note that, using the <tt class="xref py py-attr docutils literal"><span class="pre">sp_naming_function</span></tt> argument, the whole
tree structure will be initialized to use the provided parsing
function to obtain species name
information. <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.set_species_naming_function" title="ete2.PhyloNode.set_species_naming_function"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></tt></a> (present in
all tree nodes) can be used to change the behavior in a previously
loaded tree, or to set different parsing function to different parts
of the tree.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>

<span class="c"># Let&#39;s use our own leaf name parsing function to obtain species</span>
<span class="c"># names. All we need to do is create a python function that takes</span>
<span class="c"># node&#39;s name as argument and return its corresponding species name.</span>
<span class="k">def</span> <span class="nf">get_species_name</span><span class="p">(</span><span class="n">node_name_string</span><span class="p">):</span>
    <span class="c"># Species code is the first part of leaf name (separated by an</span>
    <span class="c">#  underscore character)</span>
    <span class="n">spcode</span> <span class="o">=</span> <span class="n">node_name_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c"># We could even translate the code to complete names</span>
    <span class="n">code2name</span> <span class="o">=</span> <span class="p">{</span>
      <span class="s">&quot;Dme&quot;</span><span class="p">:</span><span class="s">&quot;Drosophila melanogaster&quot;</span><span class="p">,</span>
      <span class="s">&quot;Hsa&quot;</span><span class="p">:</span><span class="s">&quot;Homo sapiens&quot;</span><span class="p">,</span>
      <span class="s">&quot;Ptr&quot;</span><span class="p">:</span><span class="s">&quot;Pan troglodytes&quot;</span><span class="p">,</span>
      <span class="s">&quot;Mms&quot;</span><span class="p">:</span><span class="s">&quot;Mus musculus&quot;</span><span class="p">,</span>
      <span class="s">&quot;Cfa&quot;</span><span class="p">:</span><span class="s">&quot;Canis familiaris&quot;</span>
      <span class="p">}</span>
    <span class="k">return</span> <span class="n">code2name</span><span class="p">[</span><span class="n">spcode</span><span class="p">]</span>

<span class="c"># Now, let&#39;s ask the tree to use our custom species naming function</span>
<span class="n">t</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="n">get_species_name</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Custom mode:&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>

<span class="c"># node: Dme_001 Species name: Drosophila melanogaster</span>
<span class="c"># node: Dme_002 Species name: Drosophila melanogaster</span>
<span class="c"># node: Hsa_001 Species name: Homo sapiens</span>
<span class="c"># node: Ptr_001 Species name: Pan troglodytes</span>
<span class="c"># node: Cfa_001 Species name: Canis familiaris</span>
<span class="c"># node: Mms_001 Species name: Mus musculus</span>
</pre></div>
</div>
</div>
<div class="section" id="manual-control-of-the-species-info">
<h3><a class="toc-backref" href="#id8">Manual control of the species info</a><a class="headerlink" href="#manual-control-of-the-species-info" title="Permalink to this headline">¶</a></h3>
<p>To disable the automatic generation of species names based on node
names, a <tt class="docutils literal"><span class="pre">None</span></tt> value can be passed to the
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.set_species_naming_function" title="ete2.PhyloNode.set_species_naming_function"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></tt></a> function. From then on,
species attribute will not be automatically updated based on the name
of nodes and it could be controlled manually.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Reads a phylogenetic tree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s">&quot;(((Hsa_001,Ptr_001),(Cfa_001,Mms_001)),(Dme_001,Dme_002));&quot;</span><span class="p">)</span>

<span class="c"># Of course, you can disable the automatic generation of species</span>
<span class="c"># names. To do so, you can set the species naming function to</span>
<span class="c"># None. This is useful to set the species names manually or for</span>
<span class="c"># reading them from a newick file. Other wise, species attribute would</span>
<span class="c"># be overwriten</span>
<span class="n">mynewick</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">(((Hsa_001[&amp;&amp;NHX:species=Human],Ptr_001[&amp;&amp;NHX:species=Chimp]),</span>
<span class="s">(Cfa_001[&amp;&amp;NHX:species=Dog],Mms_001[&amp;&amp;NHX:species=Mouse])),</span>
<span class="s">(Dme_001[&amp;&amp;NHX:species=Fly],Dme_002[&amp;&amp;NHX:species=Fly]));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">mynewick</span><span class="p">,</span> <span class="n">sp_naming_function</span><span class="o">=</span><span class="bp">None</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Disabled mode (manual set)&quot;</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
    <span class="k">print</span> <span class="s">&quot;node:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="s">&quot;Species name:&quot;</span><span class="p">,</span> <span class="n">n</span><span class="o">.</span><span class="n">species</span>

<span class="c"># node: Dme_001 Species name: Fly</span>
<span class="c"># node: Dme_002 Species name: Fly</span>
<span class="c"># node: Hsa_001 Species name: Human</span>
<span class="c"># node: Ptr_001 Species name: Chimp</span>
<span class="c"># node: Cfa_001 Species name: Dog</span>
<span class="c"># node: Mms_001 Species name: Mouse</span>
</pre></div>
</div>
<p><strong>Full Example:</strong> <a class="reference download internal" href="../_downloads/species_aware_phylogenies.py"><tt class="xref download docutils literal"><span class="pre">Species</span> <span class="pre">aware</span> <span class="pre">trees</span></tt></a>.</p>
</div>
</div>
<div class="section" id="detecting-evolutionary-events">
<h2><a class="toc-backref" href="#id9">Detecting evolutionary events</a><a class="headerlink" href="#detecting-evolutionary-events" title="Permalink to this headline">¶</a></h2>
<p>There are several ways to automatically detect duplication and
speciation nodes. ETE provides two methodologies: One implements the
algorithm described in <a class="reference external" href="http://genomebiology.com/2007/8/6/R109">Huerta-Cepas (2007)</a> and is based on the species
overlap (SO) between partitions and thus does not depend on the
availability of a species tree. The second, which requires the
comparison between the gene tree and a previously defined species
tree, implements a strict tree reconciliation algorithm (Page and
Charleston, 1997). By detecting evolutionary events, orthology and
paralogy relationships among sequences can also be inferred.  Find a
comparison of both methods in <a class="reference external" href="http://www.plosone.org/article/info:doi%2F10.1371%2Fjournal.pone.0004357">Marcet-Houben and Gabaldon (2009)</a>.</p>
<div class="section" id="species-overlap-so-algorithm">
<h3><a class="toc-backref" href="#id10">Species Overlap (SO) algorithm</a><a class="headerlink" href="#species-overlap-so-algorithm" title="Permalink to this headline">¶</a></h3>
<p>In order to apply the SO algorithm, you can use the
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_descendant_evol_events" title="ete2.PhyloNode.get_descendant_evol_events"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_descendant_evol_events()</span></tt></a> method (it will detect
all evolutionary events under the current node) or the
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_my_evol_events" title="ete2.PhyloNode.get_my_evol_events"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_my_evol_events()</span></tt></a> method (it will detect only the
evolutionary events in which current node, a leaf, is involved).</p>
<p>By default the <strong>species overlap score (SOS) threshold</strong> is set to
0.0, which means that a single species in common between two node
branches will rise a duplication event. This has been shown to perform
the best with real data, however you can adjust the threshold using
the <tt class="docutils literal"><span class="pre">sos_thr</span></tt> argument present in both methods.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Loads an example tree</span>
<span class="n">nw</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),</span>
<span class="s">(Ptr_002,(Hsa_002,Mmu_002))));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">nw</span><span class="p">)</span>
<span class="k">print</span> <span class="n">t</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#---------|                   |          \-Mms_001</span>
<span class="c">#         |          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#          \--------|                   |</span>
<span class="c">#                   |                    \-Mmu_001</span>
<span class="c">#                   |</span>
<span class="c">#                   |          /-Ptr_002</span>
<span class="c">#                    \--------|</span>
<span class="c">#                             |          /-Hsa_002</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                        \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># To obtain all the evolutionary events involving a given leaf node we</span>
<span class="c"># use get_my_evol_events method</span>
<span class="n">matches</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">)</span>
<span class="n">human_seq</span> <span class="o">=</span> <span class="n">matches</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c"># Obtains its evolutionary events</span>
<span class="n">events</span> <span class="o">=</span> <span class="n">human_seq</span><span class="o">.</span><span class="n">get_my_evol_events</span><span class="p">()</span>
<span class="c"># Print its orthology and paralogy relationships</span>
<span class="k">print</span> <span class="s">&quot;Events detected that involve Hsa_001:&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>

<span class="c"># Alternatively, you can scan the whole tree topology</span>
<span class="n">events</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_descendant_evol_events</span><span class="p">()</span>
<span class="c"># Print its orthology and paralogy relationships</span>
<span class="k">print</span> <span class="s">&quot;Events detected from the root of the tree&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">)</span>

<span class="c"># If we are only interested in the orthology and paralogy relationship</span>
<span class="c"># among a given set of species, we can filter the list of sequences</span>
<span class="c">#</span>
<span class="c"># fseqs is a function that, given a list of sequences, returns only</span>
<span class="c"># those from human and mouse</span>
<span class="n">fseqs</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">slist</span><span class="p">:</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">slist</span> <span class="k">if</span> <span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;Hsa&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;Mms&quot;</span><span class="p">)]</span>
<span class="k">print</span> <span class="s">&quot;Paralogy relationships among human and mouse&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="c"># Prints paralogy relationships considering only human and</span>
        <span class="c"># mouse. Some duplication event may not involve such species,</span>
        <span class="c"># so they will be empty</span>
        <span class="k">print</span> <span class="s">&#39;   PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> \
            <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fseqs</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">in_seqs</span><span class="p">)),</span> \
            <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span>\
            <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fseqs</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">out_seqs</span><span class="p">))</span>

<span class="c"># Note that besides the list of events returned, the detection</span>
<span class="c"># algorithm has labeled the tree nodes according with the</span>
<span class="c"># predictions. We can use such lables as normal node features.</span>
<span class="n">dups</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">evoltype</span><span class="o">=</span><span class="s">&quot;D&quot;</span><span class="p">)</span> <span class="c"># Return all duplication nodes</span>
</pre></div>
</div>
</div>
<div class="section" id="tree-reconciliation-algorithm">
<h3><a class="toc-backref" href="#id11">Tree reconciliation algorithm</a><a class="headerlink" href="#tree-reconciliation-algorithm" title="Permalink to this headline">¶</a></h3>
<p>Tree reconciliation algorithm uses a predefined species tree to infer
all the necessary genes losses that explain a given gene tree
topology. Consequently, duplication and separation nodes will strictly
follow the species tree topology.</p>
<p>To perform a tree reconciliation analysis over a given node in a
molecular phylogeny you can use the <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.reconcile" title="ete2.PhyloNode.reconcile"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.reconcile()</span></tt></a>
method, which requires a species <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> as its first
argument. Leaf node names in the the species are expected to be the
same species codes in the gene tree (see
<a class="reference internal" href="#taxonomic-info">taxonomic_info</a>). All species codes present in the
gene tree should appear in the species tree.</p>
<p>As a result, the <a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.reconcile" title="ete2.PhyloNode.reconcile"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.reconcile()</span></tt></a> method will label the
original gene tree nodes as duplication or speciation, will return the
list of inferred events, and will return a new <strong>reconcilied tree</strong>
(<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloTree" title="ete2.PhyloTree"><tt class="xref py py-class docutils literal"><span class="pre">PhyloTree</span></tt></a> instance), in which inferred gene losses are
present and labeled.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>

<span class="c"># Loads a gene tree and its corresponding species tree. Note that</span>
<span class="c"># species names in sptree are the 3 firs letters of leaf nodes in</span>
<span class="c"># genetree.</span>
<span class="n">gene_tree_nw</span> <span class="o">=</span> <span class="s">&#39;((Dme_001,Dme_002),(((Cfa_001,Mms_001),((Hsa_001,Ptr_001),Mmu_001)),(Ptr_002,(Hsa_002,Mmu_002))));&#39;</span>
<span class="n">species_tree_nw</span> <span class="o">=</span> <span class="s">&quot;((((Hsa, Ptr), Mmu), (Mms, Cfa)), Dme);&quot;</span>
<span class="n">genetree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">gene_tree_nw</span><span class="p">)</span>
<span class="n">sptree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">species_tree_nw</span><span class="p">)</span>
<span class="k">print</span> <span class="n">genetree</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#---------|                   |          \-Mms_001</span>
<span class="c">#         |          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#          \--------|                   |</span>
<span class="c">#                   |                    \-Mmu_001</span>
<span class="c">#                   |</span>
<span class="c">#                   |          /-Ptr_002</span>
<span class="c">#                    \--------|</span>
<span class="c">#                             |          /-Hsa_002</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                        \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># Let&#39;s reconcile our genetree with the species tree</span>
<span class="n">recon_tree</span><span class="p">,</span> <span class="n">events</span> <span class="o">=</span> <span class="n">genetree</span><span class="o">.</span><span class="n">reconcile</span><span class="p">(</span><span class="n">sptree</span><span class="p">)</span>
<span class="c"># a new &quot;reconcilied tree&quot; is returned. As well as the list of</span>
<span class="c"># inferred events.</span>
<span class="k">print</span> <span class="s">&quot;Orthology and Paralogy relationships:&quot;</span>
<span class="k">for</span> <span class="n">ev</span> <span class="ow">in</span> <span class="n">events</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;S&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;ORTHOLOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">inparalogs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">orthologs</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">ev</span><span class="o">.</span><span class="n">etype</span> <span class="o">==</span> <span class="s">&quot;D&quot;</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;PARALOGY RELATIONSHIP:&#39;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">inparalogs</span><span class="p">),</span> <span class="s">&quot;&lt;====&gt;&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ev</span><span class="o">.</span><span class="n">outparalogs</span><span class="p">)</span>
<span class="c"># And we can explore the resulting reconciled tree</span>
<span class="k">print</span> <span class="n">recon_tree</span>
<span class="c"># You will notice how the reconcilied tree is the same as the gene</span>
<span class="c"># tree with some added branches. They are inferred gene losses.</span>
<span class="c">#</span>
<span class="c">#</span>
<span class="c">#                    /-Dme_001</span>
<span class="c">#          /--------|</span>
<span class="c">#         |          \-Dme_002</span>
<span class="c">#         |</span>
<span class="c">#         |                              /-Cfa_001</span>
<span class="c">#         |                    /--------|</span>
<span class="c">#         |                   |          \-Mms_001</span>
<span class="c">#---------|          /--------|</span>
<span class="c">#         |         |         |                    /-Hsa_001</span>
<span class="c">#         |         |         |          /--------|</span>
<span class="c">#         |         |          \--------|          \-Ptr_001</span>
<span class="c">#         |         |                   |</span>
<span class="c">#         |         |                    \-Mmu_001</span>
<span class="c">#          \--------|</span>
<span class="c">#                   |                    /-Mms</span>
<span class="c">#                   |          /--------|</span>
<span class="c">#                   |         |          \-Cfa</span>
<span class="c">#                   |         |</span>
<span class="c">#                   |         |                              /-Hsa</span>
<span class="c">#                    \--------|                    /--------|</span>
<span class="c">#                             |          /--------|          \-Ptr_002</span>
<span class="c">#                             |         |         |</span>
<span class="c">#                             |         |          \-Mmu</span>
<span class="c">#                              \--------|</span>
<span class="c">#                                       |                    /-Ptr</span>
<span class="c">#                                       |          /--------|</span>
<span class="c">#                                        \--------|          \-Hsa_002</span>
<span class="c">#                                                 |</span>
<span class="c">#                                                  \-Mmu_002</span>
<span class="c">#</span>
<span class="c"># And we can visualize the trees using the default phylogeny</span>
<span class="c"># visualization layout</span>
<span class="n">genetree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">recon_tree</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="a-closer-look-to-the-evolutionary-event-object">
<h3><a class="toc-backref" href="#id12">A closer look to the evolutionary event object</a><a class="headerlink" href="#a-closer-look-to-the-evolutionary-event-object" title="Permalink to this headline">¶</a></h3>
<p>Both methods, species overlap and tree reconciliation, can be used to
label each tree node as a duplication or speciation event. Thus, the
<tt class="xref py py-attr docutils literal"><span class="pre">PhyloNode.evoltype</span></tt> attribute of every node will be set to one
of the following states: <tt class="docutils literal"><span class="pre">D</span></tt> (Duplication), <tt class="docutils literal"><span class="pre">S</span></tt> (Speciation) or
<tt class="docutils literal"><span class="pre">L</span></tt> gene loss.</p>
<p>Additionally, a list of all the detected events is returned. Each
event is a python object of type <a class="reference internal" href="../reference/reference_phylo.html#ete2.phylo.EvolEvent" title="ete2.phylo.EvolEvent"><tt class="xref py py-class docutils literal"><span class="pre">phylo.EvolEvent</span></tt></a>, containing
some basic information about each event ( <tt class="xref py py-attr docutils literal"><span class="pre">etype</span></tt>,
<tt class="xref py py-attr docutils literal"><span class="pre">in_seqs</span></tt>, <tt class="xref py py-attr docutils literal"><span class="pre">out_seqs</span></tt>, <tt class="xref py py-attr docutils literal"><span class="pre">node</span></tt>):</p>
<p>If an event represents a duplication, <tt class="docutils literal"><span class="pre">in_seqs</span></tt> <strong>are all
paralogous</strong> to <tt class="docutils literal"><span class="pre">out_seqs</span></tt>. Similarly, if an event represents a
speciation, <tt class="docutils literal"><span class="pre">in_seqs</span></tt> <strong>are all orthologous</strong> to <tt class="docutils literal"><span class="pre">out_seqs</span></tt>.</p>
</div>
</div>
<div class="section" id="relative-dating-phylogenetic-nodes">
<h2><a class="toc-backref" href="#id13">Relative dating phylogenetic nodes</a><a class="headerlink" href="#relative-dating-phylogenetic-nodes" title="Permalink to this headline">¶</a></h2>
<p>In molecular phylogeny, nodes can be interpreted as evolutionary
events. Therefore, they represent duplication or speciation events. In
the case of gene duplication events, nodes can also be assigned to a
certain point in a relative temporal scale. In other words, you can
obtain a relative dating of all the duplication events detected.</p>
<p>Although <strong>absolute dating is always preferred and more precise</strong>,
topological dating provides a faster approach to compare the relative
age of paralogous sequences (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/1/38.long">read this</a> for
a comparison with other methods, such as the use of synonymous
substitution rates as a proxy to the divergence time).</p>
<p>Some applications of topological dating can be found in <a class="reference external" href="http://genomebiology.com/2007/8/6/r109">Huerta-Cepas
et al, 2007</a> or, more
recently, in <a class="reference external" href="http://bib.oxfordjournals.org/content/12/5/442.abstract">Huerta-Cepas et al, 2011</a> or
<a class="reference external" href="http://www.nature.com/nature/journal/v468/n7325/full/nature09634.html">Kalinka et al, 2001</a>.</p>
<div class="section" id="implementation">
<h3><a class="toc-backref" href="#id14">Implementation</a><a class="headerlink" href="#implementation" title="Permalink to this headline">¶</a></h3>
<p>The aim of relative dating is to establish a gradient of ages among
sequences. For this, a reference species needs to be fixed, so the
gradient of ages will be referred to that referent point.</p>
<p>Thus, if our reference species is <cite>Human</cite>, we could establish the
following gradient of species:</p>
<ul class="simple">
<li>(1) Human -&gt; (2) Other Primates -&gt; (3) Mammals -&gt; (4) Vertebrates</li>
</ul>
<p>So, nodes in a tree can be assigned to one of the above categories
depending on the sequences grouped. For instance:</p>
<ul class="simple">
<li>A node with only human sequences will be mapped to (1).</li>
<li>A node with human and orangutan sequences will be mapped to (2)</li>
<li>A node with human a fish sequences will be mapped to (4)</li>
</ul>
<p>This simple calculation can be done automatically by encoding the
gradient of species ages as Python dictionary.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">relative_dist</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s">&quot;human&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="c"># human</span>
    <span class="s">&quot;chimp&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="c"># Primates non human</span>
    <span class="s">&quot;rat&quot;</span><span class="p">:</span>   <span class="mi">2</span><span class="p">,</span> <span class="c"># Mammals non primates</span>
    <span class="s">&quot;mouse&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># Mammals non primates</span>
    <span class="s">&quot;fish&quot;</span><span class="p">:</span>  <span class="mi">3</span>  <span class="c"># Vertebrates non mammals</span>
    <span class="p">}</span>
</pre></div>
</div>
<p>Once done, ETE can check the relative age of any tree node.  The
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_age" title="ete2.PhyloNode.get_age"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_age()</span></tt></a> method can be used to that purpose.</p>
<p>For example, let&#8217;s consider the following gene tree:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c">#                         /-humanA</span>
<span class="c">#                    /---|</span>
<span class="c">#                   |     \-chimpA</span>
<span class="c">#               /Dup1</span>
<span class="c">#              |    |     /-humanB</span>
<span class="c">#          /---|     \---|</span>
<span class="c">#         |    |          \-chimpB</span>
<span class="c">#     /---|    |</span>
<span class="c">#    |    |     \-mouseA</span>
<span class="c">#    |    |</span>
<span class="c">#    |     \-fish</span>
<span class="c">#-Dup3</span>
<span class="c">#    |               /-humanC</span>
<span class="c">#    |          /---|</span>
<span class="c">#    |     /---|     \-chimpC</span>
<span class="c">#    |    |    |</span>
<span class="c">#     \Dup2     \-humanD</span>
<span class="c">#         |</span>
<span class="c">#         |     /-ratC</span>
<span class="c">#          \---|</span>
<span class="c">#               \-mouseC</span>
</pre></div>
</div>
<p>the expected node dating would be:</p>
<blockquote>
<div><ul class="simple">
<li>Dup1 will be assigned to primates (most distant species is
chimp). <tt class="docutils literal"><span class="pre">Dup1.get_age(relative_distances)</span></tt> will return 1</li>
<li>Dup2 will be assigned to mammals [2] (most distant species are rat
and mouse). <tt class="docutils literal"><span class="pre">Dup2.get_age(relative_distances)</span></tt> will return 2</li>
<li>Dup3 will be assigned to mammals [3] (most distant species is
fish). <tt class="docutils literal"><span class="pre">Dup3.get_age(relative_distances)</span></tt> will return 3</li>
</ul>
</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">PhyloTree</span>
<span class="c"># Creates a gene phylogeny with several duplication events at</span>
<span class="c"># different levels. Note that we are using the default method for</span>
<span class="c"># detecting the species code of leaves (three first lettes in the node</span>
<span class="c"># name are considered the species code).</span>
<span class="n">nw</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">((Dme_001,Dme_002),(((Cfa_001,Mms_001),((((Hsa_001,Hsa_003),Ptr_001)</span>
<span class="s">,Mmu_001),((Hsa_004,Ptr_004),Mmu_004))),(Ptr_002,(Hsa_002,Mmu_002))));</span>
<span class="s">&quot;&quot;&quot;</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">nw</span><span class="p">)</span>
<span class="k">print</span> <span class="s">&quot;Original tree:&quot;</span><span class="p">,</span>
<span class="k">print</span> <span class="n">t</span>
<span class="c">#</span>
<span class="c">#             /-Dme_001</span>
<span class="c">#   /--------|</span>
<span class="c">#  |          \-Dme_002</span>
<span class="c">#  |</span>
<span class="c">#  |                              /-Cfa_001</span>
<span class="c">#  |                    /--------|</span>
<span class="c">#  |                   |          \-Mms_001</span>
<span class="c">#  |                   |</span>
<span class="c">#--|                   |                                        /-Hsa_001</span>
<span class="c">#  |                   |                              /--------|</span>
<span class="c">#  |          /--------|                    /--------|          \-Hsa_003</span>
<span class="c">#  |         |         |                   |         |</span>
<span class="c">#  |         |         |          /--------|          \-Ptr_001</span>
<span class="c">#  |         |         |         |         |</span>
<span class="c">#  |         |         |         |          \-Mmu_001</span>
<span class="c">#  |         |          \--------|</span>
<span class="c">#   \--------|                   |                    /-Hsa_004</span>
<span class="c">#            |                   |          /--------|</span>
<span class="c">#            |                    \--------|          \-Ptr_004</span>
<span class="c">#            |                             |</span>
<span class="c">#            |                              \-Mmu_004</span>
<span class="c">#            |</span>
<span class="c">#            |          /-Ptr_002</span>
<span class="c">#             \--------|</span>
<span class="c">#                      |          /-Hsa_002</span>
<span class="c">#                       \--------|</span>
<span class="c">#                                 \-Mmu_002</span>
<span class="c"># Create a dictionary with relative ages for the species present in</span>
<span class="c"># the phylogenetic tree.  Note that ages are only relative numbers to</span>
<span class="c"># define which species are older, and that different species can</span>
<span class="c"># belong to the same age.</span>
<span class="n">species2age</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s">&#39;Hsa&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="c"># Homo sapiens (Hominids)</span>
  <span class="s">&#39;Ptr&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># P. troglodytes (primates)</span>
  <span class="s">&#39;Mmu&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="c"># Macaca mulata (primates)</span>
  <span class="s">&#39;Mms&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="c"># Mus musculus (mammals)</span>
  <span class="s">&#39;Cfa&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="c"># Canis familiaris (mammals)</span>
  <span class="s">&#39;Dme&#39;</span><span class="p">:</span> <span class="mi">4</span>  <span class="c"># Drosophila melanogaster (metazoa)</span>
<span class="p">}</span>
<span class="c"># We can translate each number to its correspondig taxonomic number</span>
<span class="n">age2name</span> <span class="o">=</span> <span class="p">{</span>
  <span class="mi">1</span><span class="p">:</span><span class="s">&quot;hominids&quot;</span><span class="p">,</span>
  <span class="mi">2</span><span class="p">:</span><span class="s">&quot;primates&quot;</span><span class="p">,</span>
  <span class="mi">3</span><span class="p">:</span><span class="s">&quot;mammals&quot;</span><span class="p">,</span>
  <span class="mi">4</span><span class="p">:</span><span class="s">&quot;metazoa&quot;</span>
<span class="p">}</span>
<span class="n">event1</span><span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">(</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">,</span> <span class="s">&quot;Hsa_004&quot;</span><span class="p">)</span>
<span class="n">event2</span><span class="o">=</span><span class="n">t</span><span class="o">.</span><span class="n">get_common_ancestor</span><span class="p">(</span><span class="s">&quot;Hsa_001&quot;</span><span class="p">,</span> <span class="s">&quot;Hsa_002&quot;</span><span class="p">)</span>
<span class="k">print</span>
<span class="k">print</span> <span class="s">&quot;The duplication event leading to the human sequences Hsa_001 and &quot;</span><span class="o">+</span>\
    <span class="s">&quot;Hsa_004 is dated at: &quot;</span><span class="p">,</span> <span class="n">age2name</span><span class="p">[</span><span class="n">event1</span><span class="o">.</span><span class="n">get_age</span><span class="p">(</span><span class="n">species2age</span><span class="p">)]</span>
<span class="k">print</span> <span class="s">&quot;The duplication event leading to the human sequences Hsa_001 and &quot;</span><span class="o">+</span>\
    <span class="s">&quot;Hsa_002 is dated at: &quot;</span><span class="p">,</span> <span class="n">age2name</span><span class="p">[</span><span class="n">event2</span><span class="o">.</span><span class="n">get_age</span><span class="p">(</span><span class="n">species2age</span><span class="p">)]</span>
<span class="c"># The duplication event leading to the human sequences Hsa_001 and Hsa_004</span>
<span class="c"># is dated at:  primates</span>
<span class="c">#</span>
<span class="c"># The duplication event leading to the human sequences Hsa_001 and Hsa_002</span>
<span class="c"># is dated at:  mammals</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Note that relative distances will vary depending on your reference
species.</p>
</div>
</div>
</div>
<div class="section" id="automatic-rooting-outgroup-detection">
<h2><a class="toc-backref" href="#id15">Automatic rooting (outgroup detection)</a><a class="headerlink" href="#automatic-rooting-outgroup-detection" title="Permalink to this headline">¶</a></h2>
<p>Two methods are provided to assist in the automatic rooting of
phylogenetic trees. Since tree nodes contain relative age information
(based on the species code autodetection), the same relative age
dictionaries can be used to detect the farthest and oldest node in a
tree to given sequences.</p>
<p><a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_farthest_oldest_node" title="ete2.PhyloNode.get_farthest_oldest_node"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_farthest_oldest_node()</span></tt></a> and
<a class="reference internal" href="../reference/reference_phylo.html#ete2.PhyloNode.get_farthest_oldest_leaf" title="ete2.PhyloNode.get_farthest_oldest_leaf"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.get_farthest_oldest_leaf()</span></tt></a> can be used for that
purpose.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Phylogenetic Trees</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#linking-phylogenetic-trees-with-multiple-sequence-alignments">Linking Phylogenetic Trees with Multiple Sequence Alignments</a></li>
<li><a class="reference internal" href="#visualization-of-phylogenetic-trees">Visualization of phylogenetic trees</a></li>
<li><a class="reference internal" href="#adding-taxonomic-information">Adding taxonomic information</a><ul>
<li><a class="reference internal" href="#automatic-control-of-species-info">Automatic control of species info</a></li>
<li><a class="reference internal" href="#automatic-and-custom-control-of-the-species-info">Automatic (and custom) control of the species info</a></li>
<li><a class="reference internal" href="#manual-control-of-the-species-info">Manual control of the species info</a></li>
</ul>
</li>
<li><a class="reference internal" href="#detecting-evolutionary-events">Detecting evolutionary events</a><ul>
<li><a class="reference internal" href="#species-overlap-so-algorithm">Species Overlap (SO) algorithm</a></li>
<li><a class="reference internal" href="#tree-reconciliation-algorithm">Tree reconciliation algorithm</a></li>
<li><a class="reference internal" href="#a-closer-look-to-the-evolutionary-event-object">A closer look to the evolutionary event object</a></li>
</ul>
</li>
<li><a class="reference internal" href="#relative-dating-phylogenetic-nodes">Relative dating phylogenetic nodes</a><ul>
<li><a class="reference internal" href="#implementation">Implementation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#automatic-rooting-outgroup-detection">Automatic rooting (outgroup detection)</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_drawing.html"
                        title="previous chapter">The Programmable Tree Drawing Engine</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_clustering.html"
                        title="next chapter">Clustering Trees</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tutorial/tutorial_phylogeny.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_clustering.html" title="Clustering Trees"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_drawing.html" title="The Programmable Tree Drawing Engine"
             >previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" >The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Jaime Huerta-Cepas &amp; Toni Gabaldón.
      Last updated on Apr 09, 2012.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.2.
    </div>
  </body>
</html>