

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>The PhylomeDB API &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.1 beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="The ETE tutorial" href="index.html" />
    <link rel="next" title="Phylogenetic XML standards" href="tutorial_xml.html" />
    <link rel="prev" title="Clustering Trees" href="tutorial_clustering.html" />
 

<!-- <div class="warning" align="center"> -->
<!--   This site is still under construction and refers to -->
<!--   the new <b>ETE 2.1</b> version (temporary module -->
<!--   name: <b>ete2a1</b>). Many sections may be incomplete or contain -->
<!--   errors. -->
<!-- </div> -->


<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-783841-12']);
  _gaq.push(['_setCookiePath', '/ete2/']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>






  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_xml.html" title="Phylogenetic XML standards"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_clustering.html" title="Clustering Trees"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="the-phylomedb-api">
<h1>The PhylomeDB API<a class="headerlink" href="#the-phylomedb-api" title="Permalink to this headline">¶</a></h1>
<p>PhylomeDB is a public database for complete collections of gene phylogenies
(phylomes). It allows users to interactively explore the evolutionary history of
genes through the visualization of phylogenetic trees and multiple sequence
alignments. Moreover, phylomeDB provides genome-wide orthology and paralogy
predictions which are based on the analysis of the phylogenetic trees. The
automated pipeline used to reconstruct trees aims at providing a high-quality
phylogenetic analysis of different genomes , including Maximum Likelihood or
Bayesian tree inference, alignment trimming and evolutionary model testing.
PhylomeDB includes also a public download section with the complete set of
trees, alignments and orthology predictions.</p>
<p>ETE&#8217;s phylomeDB extension provides an access API to the main PhylomeDB database,
thus allowing to search for and fetch precomputed gene phylogenies.</p>
<div class="section" id="basis-of-the-phylomedb-api-usage">
<h2>Basis of the phylomeDB API usage<a class="headerlink" href="#basis-of-the-phylomedb-api-usage" title="Permalink to this headline">¶</a></h2>
<p>In order to explore the database resources, you have to create a connector to
the database, which will be used to query it. To do so, you must use the
<strong>PhylomeDBConnector</strong> class and specify the parameters of the DB connection.</p>
<p>The PhylomeDBConnector constructor will return a pointer to the DB that you can
use to perform queries. All methods starting by <strong>get_</strong> can be used to retrieve
information from the database. A complete list of available methods can be found
in the ETE&#8217;s programming guide (available at
<a class="reference external" href="http://ete.cgenomics.orgete.cgenomics.org">http://ete.cgenomics.orgete.cgenomics.org</a>) or explored by executing
<strong>dir(PhylomeDBConnector)</strong> in a python console.</p>
</div>
<div class="section" id="phylomedb-structure">
<h2>PhylomeDB structure<a class="headerlink" href="#phylomedb-structure" title="Permalink to this headline">¶</a></h2>
<p>A phylome includes thousands of gene trees associated to the different
genes/proteins of a given species. Thus, for example, the human phylome includes
more than 20.000 phylogenetic trees; on per human gene. Moreover, the same gene
may be associated to different trees within the same phylome differing only in
the evolutionary model that assumed to reconstruct the phylogeny.</p>
<p>Given that each phylogenetic tree was reconstructed using a a single gene as the
seed sequence to find homologous in other species, the tree takes the name from
the seed sequence.</p>
<p>You can obtain a full list of phylomes through the <strong>get_phylomes()</strong> and a full
list of seed sequence in a phylome using the <strong>get_seed_ids()</strong> method.
Phylogenetic trees within a given phylome were reconstructed in a context of a
fixed set of species. In order to obtain the list of proteomes included in a
phylome, use the** get_proteomes_in_phylome()** method. PhylomeDB uses its own
sequence identifiers, but you can use the <strong>search_id()</strong> to find a match from
an external sequence ID.</p>
<p>Each phylome is the collection of all trees associated to a given species. Thus,
the human phylome will contain thousands of phylogenetic trees. Each
gene/protein in a phylome may be associated to different trees, testing, for
example, different evolutionary models. Thus when you query the database for a
gene phylogeny you have to specify from which phylome and which specific tree.
Alternatively, you can query for the best tree in a given phylomes, which will
basically return the best likelihood tree for the queried gene/protein. The
get_tree and get_best_tree methods carry out such operations. When trees are
fetched from the phylomeDB database, the are automatically converted to the
PhyloTree class, thus allowing to operate with them as phylogenetic trees.</p>
</div>
<div class="section" id="going-phylogenomic-scale">
<h2>Going phylogenomic scale<a class="headerlink" href="#going-phylogenomic-scale" title="Permalink to this headline">¶</a></h2>
<p>Just to show you how to explore a complete phylome:</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">The PhylomeDB API</a><ul>
<li><a class="reference internal" href="#basis-of-the-phylomedb-api-usage">Basis of the phylomeDB API usage</a></li>
<li><a class="reference internal" href="#phylomedb-structure">PhylomeDB structure</a></li>
<li><a class="reference internal" href="#going-phylogenomic-scale">Going phylogenomic scale</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_clustering.html"
                        title="previous chapter">Clustering Trees</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tutorial_xml.html"
                        title="next chapter">Phylogenetic XML standards</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tutorial/tutorial_phylomedb.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tutorial_xml.html" title="Phylogenetic XML standards"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_clustering.html" title="Clustering Trees"
             >previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" >The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Jaime Huerta-Cepas &amp; Toni Gabaldón.
      Last updated on Apr 09, 2012.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.2.
    </div>
  </body>
</html>