

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PhyloTree class &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.1 beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="ETE’s Reference Guide" href="index.html" />
    <link rel="next" title="Clustering module" href="reference_clustering.html" />
    <link rel="prev" title="Supported Color Names" href="reference_svgcolors.html" />
 

<!-- <div class="warning" align="center"> -->
<!--   This site is still under construction and refers to -->
<!--   the new <b>ETE 2.1</b> version (temporary module -->
<!--   name: <b>ete2a1</b>). Many sections may be incomplete or contain -->
<!--   errors. -->
<!-- </div> -->


<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-783841-12']);
  _gaq.push(['_setCookiePath', '/ete2/']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>






  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_clustering.html" title="Clustering module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_svgcolors.html" title="Supported Color Names"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="module-ete2.phylo"></span><div class="section" id="phylotree-class">
<h1>PhyloTree class<a class="headerlink" href="#phylotree-class" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="ete2.PhyloNode">
<em class="property">class </em><tt class="descname">PhyloNode</tt><big>(</big><em>newick=None</em>, <em>alignment=None</em>, <em>alg_format='fasta'</em>, <em>sp_naming_function=&lt;function _parse_species at 0x25629b0&gt;</em>, <em>format=0</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">ete2.coretype.tree.TreeNode</span></tt></p>
<p>Extends the standard <a class="reference internal" href="reference_tree.html#ete2.TreeNode" title="ete2.TreeNode"><tt class="xref py py-class docutils literal"><span class="pre">TreeNode</span></tt></a> instance. It adds
specific attributes and methods to work with phylogentic trees.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>newick</strong> &#8211; Path to the file containing the tree or, alternatively,
the text string containing the same information.</li>
<li><strong>alignment</strong> &#8211; file containing a multiple sequence alignment.</li>
<li><strong>alg_format</strong> &#8211; &#8220;fasta&#8221;, &#8220;phylip&#8221; or &#8220;iphylip&#8221; (interleaved)</li>
<li><strong>format</strong> &#8211; <p>sub-newick format</p>
<table border="1" class="docutils">
<colgroup>
<col width="12%" />
<col width="88%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">FORMAT</th>
<th class="head">DESCRIPTION</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0</td>
<td>flexible with support values</td>
</tr>
<tr class="row-odd"><td>1</td>
<td>flexible with internal node names</td>
</tr>
<tr class="row-even"><td>2</td>
<td>all branches + leaf names + internal supports</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>all branches + all names</td>
</tr>
<tr class="row-even"><td>4</td>
<td>leaf branches + leaf names</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>internal and leaf branches + leaf names</td>
</tr>
<tr class="row-even"><td>6</td>
<td>internal branches + leaf names</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>leaf branches + all names</td>
</tr>
<tr class="row-even"><td>8</td>
<td>all names</td>
</tr>
<tr class="row-odd"><td>9</td>
<td>leaf names</td>
</tr>
<tr class="row-even"><td>100</td>
<td>topology only</td>
</tr>
</tbody>
</table>
</li>
<li><strong>sp_naming_function</strong> &#8211; Pointer to a parsing python
function that receives nodename as first argument and returns
the species name (see
<a class="reference internal" href="#ete2.PhyloNode.set_species_naming_function" title="ete2.PhyloNode.set_species_naming_function"><tt class="xref py py-func docutils literal"><span class="pre">PhyloNode.set_species_naming_function()</span></tt></a>. By default, the
3 first letter of nodes will be used as species identifiers.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a tree node object which represents the base of the tree.</p>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="ete2.PhyloNode.get_age">
<tt class="descname">get_age</tt><big>(</big><em>species2age</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_age" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_age_balanced_outgroup">
<tt class="descname">get_age_balanced_outgroup</tt><big>(</big><em>species2age</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_age_balanced_outgroup" title="Permalink to this definition">¶</a></dt>
<dd><p class="versionadded">
<span class="versionmodified">New in version 2.x.</span></p>
<p>Returns the best outgroup according to topological ages and
node sizes.</p>
<p>Currently Experimental !!</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_descendant_evol_events">
<tt class="descname">get_descendant_evol_events</tt><big>(</big><em>sos_thr=0.0</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_descendant_evol_events" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of <strong>all</strong> duplication and speciation
events detected after this node. Nodes are assumed to be
duplications when a species overlap is found between its child
linages. Method is described more detail in:</p>
<p>&#8220;The Human Phylome.&#8221; Huerta-Cepas J, Dopazo H, Dopazo J, Gabaldon
T. Genome Biol. 2007;8(6):R109.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_farthest_oldest_leaf">
<tt class="descname">get_farthest_oldest_leaf</tt><big>(</big><em>species2age</em>, <em>is_leaf_fn=None</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_farthest_oldest_leaf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the farthest oldest leaf to the current
one. It requires an species2age dictionary with the age
estimation for all species.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>is_leaf_fn</strong> (<em>None</em>) &#8211; A pointer to a function that
receives a node instance as unique argument and returns True
or False. It can be used to dynamically collapse nodes, so
they are seen as leaves.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_farthest_oldest_node">
<tt class="descname">get_farthest_oldest_node</tt><big>(</big><em>species2age</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_farthest_oldest_node" title="Permalink to this definition">¶</a></dt>
<dd><p class="versionadded">
<span class="versionmodified">New in version 2.1.</span></p>
<p>Returns the farthest oldest node (leaf or internal). The
difference with get_farthest_oldest_leaf() is that in this
function internal nodes grouping seqs from the same species
are collapsed.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_my_evol_events">
<tt class="descname">get_my_evol_events</tt><big>(</big><em>sos_thr=0.0</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_my_evol_events" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of duplication and speciation events in
which the current node has been involved. Scanned nodes are
also labeled internally as dup=True|False. You can access this
labels using the &#8216;node.dup&#8217; sintaxis.</p>
<p>Method: the algorithm scans all nodes from the given leafName to
the root. Nodes are assumed to be duplications when a species
overlap is found between its child linages. Method is described
more detail in:</p>
<p>&#8220;The Human Phylome.&#8221; Huerta-Cepas J, Dopazo H, Dopazo J, Gabaldon
T. Genome Biol. 2007;8(6):R109.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.get_species">
<tt class="descname">get_species</tt><big>(</big><big>)</big><a class="headerlink" href="#ete2.PhyloNode.get_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the set of species covered by its partition.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.is_monophyletic">
<tt class="descname">is_monophyletic</tt><big>(</big><em>species</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.is_monophyletic" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True id species names under this node are all
included in a given list or set of species names.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.iter_species">
<tt class="descname">iter_species</tt><big>(</big><big>)</big><a class="headerlink" href="#ete2.PhyloNode.iter_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an iterator over the species grouped by this node.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.link_to_alignment">
<tt class="descname">link_to_alignment</tt><big>(</big><em>alignment</em>, <em>alg_format='fasta'</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.link_to_alignment" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.reconcile">
<tt class="descname">reconcile</tt><big>(</big><em>species_tree</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.reconcile" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the reconcilied topology with the provided species
tree, and a list of evolutionary events inferred from such
reconciliation.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhyloNode.set_species_naming_function">
<tt class="descname">set_species_naming_function</tt><big>(</big><em>fn</em><big>)</big><a class="headerlink" href="#ete2.PhyloNode.set_species_naming_function" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the parsing function used to extract species name from a
node&#8217;s name.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fn</strong> &#8211; Pointer to a parsing python function that
receives nodename as first argument and returns the species
name.</td>
</tr>
</tbody>
</table>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Example of a parsing function to extract species names for</span>
<span class="c"># all nodes in a given tree.</span>
<span class="k">def</span> <span class="nf">parse_sp_name</span><span class="p">(</span><span class="n">node_name</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">node_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">tree</span><span class="o">.</span><span class="n">set_species_naming_function</span><span class="p">(</span><span class="n">parse_sp_name</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="ete2.PhyloNode.species">
<tt class="descname">species</tt><a class="headerlink" href="#ete2.PhyloNode.species" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="ete2.PhyloTree">
<tt class="descname">PhyloTree</tt><a class="headerlink" href="#ete2.PhyloTree" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#ete2.PhyloNode" title="ete2.PhyloNode"><tt class="xref py py-class docutils literal"><span class="pre">PhyloNode</span></tt></a></p>
</dd></dl>

<dl class="class">
<dt id="ete2.phylo.EvolEvent">
<em class="property">class </em><tt class="descname">EvolEvent</tt><a class="headerlink" href="#ete2.phylo.EvolEvent" title="Permalink to this definition">¶</a></dt>
<dd><p>Basic evolutionary event. It stores all the information about an
event(node) ocurred in a phylogenetic tree.</p>
<p><tt class="xref py py-attr docutils literal"><span class="pre">etype</span></tt> : <tt class="docutils literal"><span class="pre">D</span></tt> (Duplication), <tt class="docutils literal"><span class="pre">S</span></tt> (Speciation), <tt class="docutils literal"><span class="pre">L</span></tt> (gene loss),</p>
<p><tt class="xref py py-attr docutils literal"><span class="pre">in_seqs</span></tt> : the list of sequences in one side of the event.</p>
<p><tt class="xref py py-attr docutils literal"><span class="pre">out_seqs</span></tt> : the list of sequences in the other side of the event</p>
<p><tt class="xref py py-attr docutils literal"><span class="pre">node</span></tt> : link to the event node in the tree</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_svgcolors.html"
                        title="previous chapter">Supported Color Names</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_clustering.html"
                        title="next chapter">Clustering module</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/reference/reference_phylo.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_clustering.html" title="Clustering module"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_svgcolors.html" title="Supported Color Names"
             >previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" >ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Jaime Huerta-Cepas &amp; Toni Gabaldón.
      Last updated on Apr 09, 2012.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.2.
    </div>
  </body>
</html>