*** types.py.original	Wed Jan 07 01:59:55 2009
--- types.py	Wed May 19 16:30:13 2010
***************
*** 327,333 ****
      if v < 0:
          sign = 16384
          v *= -1
!     max_weight = decimal.Decimal(int(math.floor(math.log(v) / math.log(10000))))
      weight = max_weight
      digits = []
      while v != 0:
--- 327,333 ----
      if v < 0:
          sign = 16384
          v *= -1
!     max_weight = decimal.Decimal(int(math.floor(math.log(v or 1) / math.log(10000))))
      weight = max_weight
      digits = []
      while v != 0:
***************
*** 335,341 ****
          v = v - (digit * (10000 ** weight))
          weight -= 1
          digits.append(digit)
!     retval = struct.pack("!hhhh", len(digits), max_weight, sign, 0)
      retval += struct.pack("!" + ("h" * len(digits)), *digits)
      return retval
  
--- 335,341 ----
          v = v - (digit * (10000 ** weight))
          weight -= 1
          digits.append(digit)
!     retval = struct.pack("!hhhh", len(digits), int(max_weight), sign, 0)
      retval += struct.pack("!" + ("h" * len(digits)), *digits)
      return retval
  
