index 9afb6ad..cfd6b4c 100644
--- a/plone/subrequest/__init__.py
+++ b/plone/subrequest/__init__.py
@@ -61,7 +61,10 @@ def subrequest(url, root=None, stdout=None):
         path = urljoin(here, path)
         path = normpath(path)
     request = parent_request.clone()
-    request.other.update(parent_request.other)
+    for name in ['LANGUAGE', 'LANGUAGE_TOOL', 'AUTHENTICATION_PATH', 'AUTHENTICATED_USER']:
+        parent_value = parent_request.other.get(name, None)
+        if parent_value is not None:
+            request.other[name] = parent_value
     request['PARENT_REQUEST'] = parent_request
     alsoProvides(request, ISubRequest)
     try:
