

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PhylomeDB3 Connector &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.1 beta',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="ETE’s Reference Guide" href="index.html" />
    <link rel="next" title="Seqgroup class" href="reference_seqgroup.html" />
    <link rel="prev" title="Phyloxml Module" href="reference_phyloxml.html" />
 

<!-- <div class="warning" align="center"> -->
<!--   This site is still under construction and refers to -->
<!--   the new <b>ETE 2.1</b> version (temporary module -->
<!--   name: <b>ete2a1</b>). Many sections may be incomplete or contain -->
<!--   errors. -->
<!-- </div> -->


<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-783841-4']);
  _gaq.push(['_setCookiePath', '/ete2/']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>






  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_seqgroup.html" title="Seqgroup class"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="reference_phyloxml.html" title="Phyloxml Module"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="phylomedb3-connector">
<h1>PhylomeDB3 Connector<a class="headerlink" href="#phylomedb3-connector" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="ete2.PhylomeDB3Connector">
<em class="property">class </em><tt class="descname">PhylomeDB3Connector</tt><big>(</big><em>host='84.88.66.245'</em>, <em>db='phylomedb_3'</em>, <em>user='public'</em>, <em>passwd='public'</em>, <em>port=3306</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a connector to a phylomeDB3 database.</p>
<blockquote>
<div>db: database name in the host server.
host: hostname in which phylomeDB is hosted.
user: username to the database.
port: port used to connect database.
passwd: password to connect database.</div></blockquote>
<blockquote>
<div>An object whose methods can be used to query the database.</div></blockquote>
<dl class="method">
<dt id="ete2.PhylomeDB3Connector.count_algs">
<tt class="descname">count_algs</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.count_algs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns how many alignments are for a given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.count_trees">
<tt class="descname">count_trees</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.count_trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Retuns the frequency of each evolutionary method in the input phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_algs">
<tt class="descname">get_algs</tt><big>(</big><em>id</em>, <em>phylome_id</em>, <em>raw_alg=True</em>, <em>clean_alg=True</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_algs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the either the clean, the raw or both alignments for the input
phylomeDB ID in the input phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_all_isoforms">
<tt class="descname">get_all_isoforms</tt><big>(</big><em>id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_all_isoforms" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all the isoforms registered for the input phylomeDB ID</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_available_trees_by_phylome">
<tt class="descname">get_available_trees_by_phylome</tt><big>(</big><em>id</em>, <em>collateral=True</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_available_trees_by_phylome" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns information about which methods have been used to reconstruct
every tree for a given phylomeDB ID grouped by phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_best_tree">
<tt class="descname">get_best_tree</tt><big>(</big><em>id</em>, <em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_best_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>return a tree for input id in the given phylome for the best fitting
evolutionary model in terms of LK</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_clean_alg">
<tt class="descname">get_clean_alg</tt><big>(</big><em>id</em>, <em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_clean_alg" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the raw alignment for the input phylomeDB ID in the given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_collateral_seeds">
<tt class="descname">get_collateral_seeds</tt><big>(</big><em>protid</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_collateral_seeds" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the trees where the protid is presented as part of the homolog
sequences to the seed protein</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_external_ids">
<tt class="descname">get_external_ids</tt><big>(</big><em>ids</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_external_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all the external IDs registered in the &#8216;external_id&#8217; table that
are associated to the input phylomeDB IDs</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_genome_ids">
<tt class="descname">get_genome_ids</tt><big>(</big><em>taxid</em>, <em>version</em>, <em>filter_isoforms=True</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_genome_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the phylomeDB IDs for a given genome in the database filtering
out, or not, the different isoforms for each ID</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_genome_info">
<tt class="descname">get_genome_info</tt><big>(</big><em>genome</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_genome_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all available information about a registered genome/proteome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_genomes">
<tt class="descname">get_genomes</tt><big>(</big><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_genomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all current available genomes/proteomes</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_genomes_by_species">
<tt class="descname">get_genomes_by_species</tt><big>(</big><em>taxid</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_genomes_by_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all the proteomes/genomes registered for the input taxaid code</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_go_ids">
<tt class="descname">get_go_ids</tt><big>(</big><em>ids</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_go_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all available GO Terms associated to the input phylomeDB IDs</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_id_by_external">
<tt class="descname">get_id_by_external</tt><big>(</big><em>external</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_id_by_external" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the protein id associated to a given external id</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_id_translations">
<tt class="descname">get_id_translations</tt><big>(</big><em>id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_id_translations" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all the registered translations of a given phylomeDB ID</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_info_homologous_seqs">
<tt class="descname">get_info_homologous_seqs</tt><big>(</big><em>protid</em>, <em>phylome_id</em>, <em>tree=None</em>, <em>tree_method=False</em>, <em>sequence=False</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_info_homologous_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all the available information for a given set of homologous
sequences extracted from a tree from a given phylome.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_longest_isoform">
<tt class="descname">get_longest_isoform</tt><big>(</big><em>id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_longest_isoform" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the longest isoform for a given phylomeDB ID</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_new_phylomedb_id">
<tt class="descname">get_new_phylomedb_id</tt><big>(</big><em>old_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_new_phylomedb_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the conversion between an old phylomeDB ID and a new one</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_old_phylomedb_ids">
<tt class="descname">get_old_phylomedb_ids</tt><big>(</big><em>ids</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_old_phylomedb_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all old phylomeDB IDs associated to each of the input phylomeDB
IDs</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylome_algs">
<tt class="descname">get_phylome_algs</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylome_algs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all alignments available for a given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylome_info">
<tt class="descname">get_phylome_info</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylome_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns available information on a given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylome_seed_ids">
<tt class="descname">get_phylome_seed_ids</tt><big>(</big><em>phylome_id</em>, <em>filter_isoforms=True</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylome_seed_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the seed phylomeDB IDs for a given phylome being possible to
filter out the longest isoforms</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylome_seed_ids_info">
<tt class="descname">get_phylome_seed_ids_info</tt><big>(</big><em>phylome_id</em>, <em>start=0</em>, <em>offset=None</em>, <em>filter_isoforms=False</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylome_seed_ids_info" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylome_trees">
<tt class="descname">get_phylome_trees</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylome_trees" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all trees available for a given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylomes">
<tt class="descname">get_phylomes</tt><big>(</big><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all current available phylomes</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_phylomes_for_seed_ids">
<tt class="descname">get_phylomes_for_seed_ids</tt><big>(</big><em>ids</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_phylomes_for_seed_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a list of phylomeDB IDs, return in which phylomes these IDs have
been used as a seed</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_prot_gene_names">
<tt class="descname">get_prot_gene_names</tt><big>(</big><em>ids</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_prot_gene_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all possible protein and gene names associated to the input
phylomeDB IDs</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_proteomes_in_phylome">
<tt class="descname">get_proteomes_in_phylome</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_proteomes_in_phylome" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of proteomes associated to a given phylome_id</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_raw_alg">
<tt class="descname">get_raw_alg</tt><big>(</big><em>id</em>, <em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_raw_alg" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the raw alignment for the input phylomeDB ID in the given phylome</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_seq_info_in_tree">
<tt class="descname">get_seq_info_in_tree</tt><big>(</big><em>id</em>, <em>phylome_id</em>, <em>method=None</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_seq_info_in_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all the available information for each sequence from tree/s
asociated to a tuple (protein, phylome) identifiers.</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_seq_info_msf">
<tt class="descname">get_seq_info_msf</tt><big>(</big><em>id</em>, <em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_seq_info_msf" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all available information for the homologous sequences to the
input phylomeDB ID in the input phylome using the best tree to compute
the set of homologous sequences</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_seqid_info">
<tt class="descname">get_seqid_info</tt><big>(</big><em>id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_seqid_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns available information about a given protid</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_seqs_in_genome">
<tt class="descname">get_seqs_in_genome</tt><big>(</big><em>taxid</em>, <em>version</em>, <em>filter_isoforms=True</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_seqs_in_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all sequences of a given proteome, filtering the</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_species">
<tt class="descname">get_species</tt><big>(</big><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all current registered species in the database</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_species_in_phylome">
<tt class="descname">get_species_in_phylome</tt><big>(</big><em>phylome_id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_species_in_phylome" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of proteomes associated to a given phylome_id</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_species_info">
<tt class="descname">get_species_info</tt><big>(</big><em>taxid=None</em>, <em>code=None</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_species_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all information on a given species/code</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.get_tree">
<tt class="descname">get_tree</tt><big>(</big><em>id</em>, <em>phylome_id</em>, <em>method=None</em>, <em>best_tree=False</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.get_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Depending in the input parameters select either
.- a tree with the best evolutionary model in terms of LK (best_tree)
.- a tree reconstructed using a specific model (method)
.- all available model/trees for the tuple (phylomeDB ID, phylome ID)</p>
</dd></dl>

<dl class="method">
<dt id="ete2.PhylomeDB3Connector.search_id">
<tt class="descname">search_id</tt><big>(</big><em>id</em><big>)</big><a class="headerlink" href="#ete2.PhylomeDB3Connector.search_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of the longest isoforms for each proteome where the ID is
already registered. The ID can be a current phylomeDB ID version, former
phylomeDB ID or an external ID.</p>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h4>Previous topic</h4>
  <p class="topless"><a href="reference_phyloxml.html"
                        title="previous chapter">Phyloxml Module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="reference_seqgroup.html"
                        title="next chapter">Seqgroup class</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/reference/reference_phylomedb.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="reference_seqgroup.html" title="Seqgroup class"
             >next</a> |</li>
        <li class="right" >
          <a href="reference_phyloxml.html" title="Phyloxml Module"
             >previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" >ETE&#8217;s Reference Guide</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Jaime Huerta-Cepas &amp; Toni Gabaldón.
      Last updated on Dec 19, 2011.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.2.
    </div>
  </body>
</html>